src/boilerplate.rs

```diff
@@ -0,0 +1,236 @@
+                let borrowed_contents = contents.borrow();
+                let text = borrowed_contents.trim();
+                if !text.starts_with("EXAMPLE ") {
+                    return;
+                }
+                const PRE: LocalName = local_name!("pre");
+                const CODE: LocalName = local_name!("code");
+                let has_suitable_parent = node.parent_node().map_or(false, |p| {
+                    p.is_html_element(&PRE)
+                        || (p.is_html_element(&CODE)
+                            && p.parent_node().map_or(false, |p2| p2.is_html_element(&PRE)))
+                });
+                if has_suitable_parent {
+                    let path = Path::new(text[8..].trim());
+                    if is_safe_path(path) {
+                        let file_contents = read_to_str_tendril(self.example_path.join(path));
+                        self.edits
+                            .push(Edit::ReplaceText(node.clone(), file_contents))
+                    }
+                }
+            }
+            _ => (),
+        }
+    }
+
+    /// Applies the required replacements, in order.
+    pub async fn apply(self) -> io::Result<()> {
+        for edit in self.edits {
+            match edit {
+                // When parsing HTML, we need the context it's in so that the
+                // context-sensitive parsing behavior works correctly.
+                Edit::ReplaceHTML(node, replacement) => {
+                    let context = match node.parent_node() {
+                        Some(n) => n,
+                        _ => continue,
+                    };
+                    let file: File = replacement.await??;
+                    let new_children = parser::parse_fragment_async(file, &context).await?;
+                    node.replace_with(new_children);
+                }
+                Edit::ReplaceAttr(element, ref attr, replacement) => {
+                    element.set_attribute(attr, replacement.await??.into());
+                }
+                Edit::ReplaceText(element, replacement) => match element.data {
+                    NodeData::Text { ref contents } => {
+                        contents.replace(replacement.await??.into());
+                    }
+                    _ => panic!("not text"),
+                },
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Check that a path is safe to open, even if the source is potentially untrusted.
+fn is_safe_path(path: &Path) -> bool {
+    use std::path::Component;
+    path.components()
+        .all(|c| matches!(c, Component::Normal(_) | Component::CurDir))
+}
+
+/// In a spawned task, read to a string, then move it to a tendril.
+fn read_to_str_tendril(path: impl AsRef<Path>) -> JoinHandle<io::Result<SendStrTendril>> {
+    let path = path.as_ref().to_owned();
+    tokio::spawn(async move {
+        let string = tokio::fs::read_to_string(path).await?;
+        Ok(StrTendril::from(string).into_send())
+    })
+}
```

Comment:
Should these two functions move to a `utils.rs`, maybe?

Here and elsewhere, I'm not really sure; I can understand keeping things more encapsulated/local, but it would also be a shame if someone in the future realized they needed something like this for another processor and didn't know about the helpers.

Reply:
Moved to an `io_utils.rs`.